### PR TITLE
add SwitchU

### PR DESCRIPTION
**Pull request by dwincort**
In principle, variant options much up with Haskell constructors, while record fields match up with Haskell fields.  Constructors start with uppercase letters, and fields start with lowercase letters.  However, row-types has a `switch` function that allows one to combine a `Var` with a `Rec`, matching up the labels.  This is a little strange, since presumably, the labels for the `Var` have different cases than those for the `Rec`.

This PR introduces a variant of `switch` called `switchU` that matches labels so long as the first letter of the `Rec` label is the same as the lowercase version of the first letter of the `Var` label and the rest of the letters are the same.

The code is inspired by (and copied from) https://kcsongor.github.io/symbol-parsing-haskell/.

### CODE REVIEW

File: Data/Row/Switch.hs [hunk 1]
```diff
@@ -43,3 +45,34 @@ instance (KnownSymbol l, Switch (R v) (R r) b)
     Left x  -> (r .! l) x
     Right v -> switch v (unsafeRemove l r)
     where l = Label @l
+
+-- | A 'Var' and a 'Rec' can combine if their rows line up properly.
+-- For the "U" variant, we match uppercase variant names with lowercase
+-- record names.  In other words, the variant name "Option1" would match with
+-- the record name "option1".
+class SwitchU (v :: Row *) (r :: Row *) x | v x -> r, r x -> v where
```
[the remark this answers is not in the document]
Comment (dwincort) [reply Data/Row/Switch.hs]: If so, I don't know how to do it :/.

**Comment by dwincort** [comment]
Now that I see that this approach doesn't work for unicode characters, I'm a little disappointed.  I think this PR should remain in limbo (i.e., unmerged) for now as I don't think anyone is directly demanding it, and I'd like to find a better solution.

Ideally, these kind of `Symbol` operations would be handled in GHC directly, or, at the very least, not in row-types.  Maybe I can make a library just for `Symbol` operations and plan for GHC to eventually make it obsolete.

**Comment by shlevy** [comment]
@dwincort Presumably the rules here are already encoded in GHC... Why not just open the PR for a symbol primitive there?

File: Data/Row/Internal.hs [hunk 2]
```diff
@@ -528,3 +535,176 @@ type a <=.? b = (CmpSymbol a b == 'LT)
...
 infix 4 ≈
 type a ≈ b = a ~ b
+
+-- | Only used at the type level
+data Tree a
+  = Leaf
+  | Node (Tree a) a (Tree a)
+
+-- | A tree for finding the first character of a symbol
+type Chars
```
[the remark this answers is not in the document]
Comment (strake) [reply Data/Row/Internal.hs]: Yeah, this needs more comments for sure.
What will it do for non-ASCII symbols?
Can it be generated, with either a build script or Template Haskell?

File: Data/Row/Internal.hs [hunk 3]
```diff
@@ -528,3 +535,176 @@ type a <=.? b = (CmpSymbol a b == 'LT)
...
+  Lowercase "R" = "r"
+  Lowercase "S" = "s"
+  Lowercase "T" = "t"
+  Lowercase "U" = "u"
+  Lowercase "V" = "v"
+  Lowercase "W" = "w"
+  Lowercase "X" = "x"
+  Lowercase "Y" = "y"
+  Lowercase "Z" = "z"
+  Lowercase  x  =  x
```
[the remark this answers is not in the document]
Comment (strake) [reply Data/Row/Internal.hs]: Is it possible to generate this also?

File: Data/Row/Switch.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (strake) [reply Data/Row/Switch.hs]: Fair enough — i hoped we could merge `Switch` and `SwitchU`, but given how the latter matches the syntax of Haskell itself, i think it's fine to have a special case for it.

**Comment by dwincort** [comment]
Turns out kcsongor already made a package for symbol manipulation, so I'm relying on that instead of writing it  out manually in row-types.

@shlevy It looks a little complicated to make the change directly in GHC.  On the one hand, I think it would be much easier because you can treat the symbols more like strings, but on the other, I think it's some weird built-in stuff that I'm not entirely sure how to mess with.  I think I'm going to punt for now.

**Comment by dwincort** [comment]
I don't think this is necessary.

**Pull request closed without merging**